cdk/lib/admin-console.ts

```diff
@@ -5,25 +5,86 @@ import { GuStack } from '@guardian/cdk/lib/constructs/core';
 import { GuCname } from '@guardian/cdk/lib/constructs/dns';
 import {
   GuAllowPolicy,
+  GuDynamoDBReadPolicy,
+  GuDynamoDBWritePolicy,
   GuGetS3ObjectsPolicy,
   GuPutS3ObjectsPolicy,
 } from '@guardian/cdk/lib/constructs/iam';
-import type { App } from 'aws-cdk-lib';
-import { Duration } from 'aws-cdk-lib';
+import type { App, CfnElement } from 'aws-cdk-lib';
+import { Duration, RemovalPolicy } from 'aws-cdk-lib';
+import { AttributeType, ProjectionType, Table } from 'aws-cdk-lib/aws-dynamodb';
 import { InstanceClass, InstanceSize, InstanceType } from 'aws-cdk-lib/aws-ec2';
 
 export interface AdminConsoleProps extends GuStackProps {
   domainName: string;
 }
 
 export class AdminConsole extends GuStack {
+  // Build a dynamodb table to store tests for the given channel
+  buildTestsTable(channel: string): Table {
+    const id = `${channel[0].toUpperCase() + channel.slice(1)}TestsDynamoTable`;
+
+    const table = new Table(this, id, {
+      tableName: `${channel}-tests-${this.stage}`,
+      removalPolicy: RemovalPolicy.RETAIN,
+      readCapacity: 4,
+      writeCapacity: 4,
+      partitionKey: {
+        name: 'name',
+        type: AttributeType.STRING,
+      },
+    });
+
+    // Add an index for querying by status (LIVE/DRAFT/ARCHIVED)
+    table.addGlobalSecondaryIndex({
+      indexName: 'status-index',
+      partitionKey: {
+        name: 'status',
+        type: AttributeType.STRING,
+      },
+      readCapacity: 4,
+      writeCapacity: 4,
+      projectionType: ProjectionType.ALL,
+    });
+
+    // Give it a better name
+    const defaultChild = table.node.defaultChild as unknown as CfnElement;
+    defaultChild.overrideLogicalId(id);
+
+    return table;
+  }
+
+  buildDynamoPolicies(tables: Table[]): GuAllowPolicy[] {
+    const dynamoReadPolicies = tables.map(
+      (table) =>
+        new GuDynamoDBReadPolicy(this, `DynamoRead-${table.node.id}`, {
+          tableName: table.tableName,
+        }),
+    );
+    const dynamoWritePolicies = tables.map(
+      (table) =>
+        new GuDynamoDBWritePolicy(this, `DynamoWrite-${table.node.id}`, {
+          tableName: table.tableName,
+        }),
+    );
+
+    return [...dynamoReadPolicies, ...dynamoWritePolicies];
+  }
+
   constructor(scope: App, id: string, props: AdminConsoleProps) {
     super(scope, id, props);
 
     const { domainName } = props;
 
     const app = 'admin-console';
 
+    const dynamoTables = [
+      this.buildTestsTable('header'),
+      this.buildTestsTable('epic'),
+      this.buildTestsTable('banner'),
+    ];
+    const dynamoPolicies = this.buildDynamoPolicies(dynamoTables);
```

Comment:
this is so much nicer than having to write things out manually in cloudformation!

Reply:
Agreed!